README.rst
Django Privacy
==============

This is a generic app for assigning privacy related settings to whole objects
or to single fields.

If you add e.g. user profiles to your app, you might want to let the users
decide where and how to publish their content.
Some might want to hide everything from public, some might want to show private
updates to friends or colleagues only.

We define clearance levels to allow access or not. The lowest level is the
least secure. E.g.:

* Level 1: Everyone can see the content.
* Level 2: Only connected users can see the content.
* Level 3: Only the user herself can see the content.

This app provides the backend for custom functions to (not) display content.

Prerequisites
-------------

You need at least the following packages in your virtualenv:

* Django >= 1.5
* django-hvad >= 0.3
* django-libs >= 1.27.1
* South


Installation
------------

To get the latest stable release from PyPi::

    $ pip install django-privacy

To get the latest commit from GitHub::

    $ pip install -e git://github.com/bitmazk/django-privacy.git#egg=privacy

Add the app to your ``INSTALLED_APPS``::

    INSTALLED_APPS = [
        ...
        'privacy',
    ]

Run the south migrations to create the app's database tables::

    $ ./manage.py migrate privacy

Settings
--------

PRIVACY_LEVEL_ORDERING
++++++++++++++++++++++

Default: ['clearance_level']

Ordering of the privacy levels. Use ``['-clearance_level']`` to reverse the
ordering or e.g. ``['name']`` to sort by level names.

PRIVACY_CLEARANCE_LEVEL_FUNCTION (mandatory)
++++++++++++++++++++++++++++++++++++++++++++

Default: None

Custom function to return a privacy level instance for the owner/requester
relation. E.g. ``'myproject.privacy_settings.get_clearance_level'``

You can find an example in the ``test_app`` of this repository.

PRIVACY_DEFAULT_CLEARANCE_LEVEL
+++++++++++++++++++++++++++++++

Default: 1

Default clearance level if a field has no privacy setting assigned.

Usage
-----

Template Tags
+++++++++++++

``is_access_allowed``

This assignment tag will check the relation of an object owner to the requester
and will return a ``True`` or ``False``. You can check the level for the whole
object or for a field of the object::

    {% load privacy_tags %}
    {% is_access_allowed object.user request.user object 'first_name' as access_allowed %}
    {% if access_allowed %}
        {{ object.first_name }}
    {% endif %}

--------------------------------------------------------------------------------

``render_privacy_level_field``

This inclusion tag will generate a privacy level next to your standard form
field::

    {% load privacy_tags %}
    {% render_privacy_level_field form.instance %}
    {% for field in form %}
        {{ field }} {% render_privacy_level_field form.instance field.field.name %}
    {% endfor %}

For carefree update forms use it with our ``PrivacyFormMixin`` (see below).


--------------------------------------------------------------------------------

``get_privacy_setting``

This assignment tag will return the current privacy setting::

    {% load privacy_tags %}
    {% get_privacy_setting form.instance 'name' as current_setting %}

The tag can be used with a field name or without to get the setting of the
whole object.


Form Mixin
++++++++++

``PrivacyFormMixin``

This form mixin handles privacy related form data (in case you use our
``render_privacy_level_field`` template tag). Just add the mixin class to your
form and privacy settings will be saved::

    from django import forms
    from privacy.forms import PrivacyFormMixin
    class MyModelForm(PrivacyFormMixin, forms.ModelForm):
        class Meta:
            model = MyModel
            fields = ('field1', 'field2')


Queryset helper
+++++++++++++++

If you want to filter querysets outside of templates (to keep pagination alive
or to use custom model managers), you can use the following helper:

``filter_privacy_level``

Just pass a ``queryset`` and a ``clearance_level``. You can also filter for
matching levels, just use ``exact``. An example::

    class MyListView(ListView):
        model = MyModel

        def get_queryset(self):
            qs = super(MyListView, self).get_queryset()
            clearance_level = get_clearance_level(self.owner, self.request.user)
            return filter_privacy_level(qs, clearance_level, self.request.GET.get('exact'))


Roadmap
-------

See the issue tracker for current and upcoming features.


CHANGELOG.txt
=== ongoing ===

=== 0.2 ===

- added get_privacy_setting filter

=== 0.1.2 ===

- Fixed form mixin to handle unsaved form instances

=== 0.1.1 ===

- Fixed helper

=== 0.1 ===

- Initial commit


privacy/tests/urls.py
"""
This ``urls.py`` is only used when running the tests via ``runtests.py``.
As you know, every app must be hooked into yout main ``urls.py`` so that
you can actually reach the app's views (provided it has any views, of course).

"""
from django.conf.urls.defaults import include, patterns, url
from django.contrib import admin
from django.views.generic import UpdateView

from test_app.forms import DummyProfileModelForm
from test_app.models import DummyProfileModel


admin.autodiscover()


urlpatterns = patterns(
    '',
    url(r'^admin/', include(admin.site.urls)),
    url(r'(?P<pk>\d+)/',
        UpdateView.as_view(
            model=DummyProfileModel, form_class=DummyProfileModelForm),
        name='dummy_profile_update')
)


privacy/tests/factories.py
"""Factories for the ``privacy`` app."""
import factory

from django_libs.tests.factories import HvadFactoryMixin

from ..models import PrivacyLevel, PrivacySetting
from test_app.factories import DummyModelFactory


class PrivacyLevelFactory(HvadFactoryMixin, factory.DjangoModelFactory):
    FACTORY_FOR = PrivacyLevel

    language_code = 'en'
    name = factory.Sequence(lambda x: 'A name {0}'.format(x))
    clearance_level = factory.Sequence(lambda x: x)


class PrivacySettingFactory(factory.DjangoModelFactory):
    FACTORY_FOR = PrivacySetting

    level = factory.SubFactory(PrivacyLevelFactory)
    content_object = factory.SubFactory(DummyModelFactory)


privacy/tests/tags_tests.py
"""Tests for the template tags of the ``privacy`` app."""
from django.template import Context, Template
from django.test import TestCase

from ..templatetags.privacy_tags import is_access_allowed, get_privacy_setting
from .factories import PrivacyLevelFactory, PrivacySettingFactory
from test_app.factories import DummyProfileModelFactory, DummyModelFactory


class GetPrivacySettingTestCase(TestCase):
    """Tests for the ``get_privacy_setting`` filter tag."""
    longMessage = True

    def setUp(self):
        self.obj = DummyModelFactory()
        self.level_1 = PrivacyLevelFactory(clearance_level=1)
        self.level_2 = PrivacyLevelFactory(clearance_level=2)

    def test_tag(self):
        self.assertIsNone(get_privacy_setting(self.obj), msg=(
            'Should return None, if there\'s no setting set.'))
        PrivacySettingFactory(level=self.level_1, content_object=self.obj)
        self.assertEqual(
            get_privacy_setting(self.obj).level.clearance_level, 1, msg=(
                'Should return the privacy setting.'))


class RenderPrivacyLevelFieldTestCase(TestCase):
    """Tests for the ``render_privacy_level_field`` inclusion tag."""
    longMessage = True

    def setUp(self):
        self.obj = DummyModelFactory()
        self.level_1 = PrivacyLevelFactory(clearance_level=1)
        self.level_2 = PrivacyLevelFactory(clearance_level=2)

    def test_tag(self):
        t = Template(
            '{% load privacy_tags %}{% render_privacy_level_field obj %}')
        c = Context({'obj': self.obj})
        self.assertIn('name="privacy_for_instance"', t.render(c))
        t = Template('{% load privacy_tags %}{% render_privacy_level_field obj'
                     ' "name" %}')
        self.assertIn('name="privacy_name"', t.render(c))


class IsAccessAllowedTestCase(TestCase):
    """Tests for the ``is_access_allowed`` assignment tag."""
    longMessage = True

    def setUp(self):
        self.level_1 = PrivacyLevelFactory(clearance_level=1)
        self.level_2 = PrivacyLevelFactory(clearance_level=2)
        self.level_3 = PrivacyLevelFactory(clearance_level=3)
        self.level_4 = PrivacyLevelFactory(clearance_level=4)
        self.owner = DummyProfileModelFactory()
        self.requester = DummyProfileModelFactory()
        self.obj = DummyModelFactory()

    def test_tag(self):
        # No level
        self.assertTrue(
            is_access_allowed(self.owner, self.requester, self.obj),
            msg=('Access should be allowed, if it\'s an object without privacy'
                 ' setting.'))

        # Level 1
        setting = PrivacySettingFactory(
            level=self.level_1, content_object=self.obj)
        self.assertTrue(
            is_access_allowed(self.owner, self.requester, self.obj),
            msg=('Access should be allowed, if it\'s a public object.'))

        # Level 2
        setting.level = self.level_2
        setting.save()
        self.assertFalse(
            is_access_allowed(self.owner, self.requester, self.obj),
            msg=('Access should not be allowed, if the requester is not at'
                 ' least a friend of a friend.'))
        friend = DummyProfileModelFactory()
        self.owner.friends.add(friend.user)
        self.requester.friends.add(friend.user)
        self.assertTrue(
            is_access_allowed(self.owner, self.requester, self.obj),
            msg=('Access should be allowed, if the requester is at least a'
                 ' friend of a friend.'))

        # Level 3
        setting.level = self.level_3
        setting.save()
        self.assertFalse(
            is_access_allowed(self.owner, self.requester, self.obj),
            msg=('Access should not be allowed, if the requester is not at'
                 ' least a friend.'))
        self.owner.friends.add(self.requester.user)
        self.assertTrue(
            is_access_allowed(self.owner, self.requester, self.obj),
            msg=('Access should be allowed, if the requester is at least a'
                 ' friend.'))

        # Level 4
        setting.level = self.level_4
        setting.save()
        self.assertFalse(
            is_access_allowed(self.owner, self.requester, self.obj),
            msg=('Access should only be allowed, if the requester is also the'
                 ' owner.'))
        self.assertTrue(
            is_access_allowed(self.owner, self.owner, self.obj),
            msg=('Access should only be allowed, if the requester is also the'
                 ' owner.'))

    def test_tag_with_field_name(self):
        # No level
        self.assertTrue(
            is_access_allowed(self.owner, self.requester, self.obj, 'name'),
            msg=('Access should be allowed, if it\'s an object without privacy'
                 ' setting.'))

        # Level 1
        setting = PrivacySettingFactory(
            level=self.level_1, content_object=self.obj, field_name='name')
        self.assertTrue(
            is_access_allowed(self.owner, self.requester, self.obj, 'name'),
            msg=('Access should be allowed, if it\'s a public object.'))

        # Level 2
        setting.level = self.level_2
        setting.save()
        self.assertFalse(
            is_access_allowed(self.owner, self.requester, self.obj, 'name'),
            msg=('Access should not be allowed, if the requester is not at'
                 ' least a friend of a friend.'))
        friend = DummyProfileModelFactory()
        self.owner.friends.add(friend.user)
        self.requester.friends.add(friend.user)
        self.assertTrue(
            is_access_allowed(self.owner, self.requester, self.obj, 'name'),
            msg=('Access should be allowed, if the requester is at least a'
                 ' friend of a friend.'))


privacy/forms.py
"""Forms for the ``privacy`` app."""
from django.contrib.contenttypes.models import ContentType

from .models import PrivacyLevel, PrivacySetting


class PrivacyFormMixin(object):
    """Form mixin for forms with privacy level fields."""
    def save(self, *args, **kwargs):
        instance = super(PrivacyFormMixin, self).save(*args, **kwargs)
        for key in self.data.keys():
            if not key.startswith('privacy_'):
                continue
            field_name = key.replace('privacy_', '')
            if field_name == 'for_instance':
                field_name = ''
            ctype = ContentType.objects.get_for_model(self.instance)
            kwargs = {
                'content_type': ctype,
                'object_id': self.instance.pk,
                'field_name': field_name,
            }
            level = PrivacyLevel.objects.get(clearance_level=self.data[key])
            try:
                privacy_setting = PrivacySetting.objects.get(**kwargs)
            except PrivacySetting.DoesNotExist:
                kwargs.update({'level': level})
                privacy_setting = PrivacySetting.objects.create(**kwargs)
            else:
                privacy_setting.level = level
                privacy_setting.save()
        return instance


privacy/models.py
"""Models for the ``privacy`` app."""
from django.conf import settings
from django.db import models
from django.contrib.contenttypes.models import ContentType
from django.contrib.contenttypes import generic
from django.utils.translation import ugettext_lazy as _

from hvad.models import TranslatableModel, TranslatedFields


def filter_privacy_level(qs, clearance_level, exact=False):
    """
    Function to exclude objects from a queryset, which got a higher clearance
    level than the wanted maximum clearance level.

    :qs: Django queryset.
    :clearance_level: Minimum clearance level.
    :exact: Boolean to check for the exact clearance level.

    """
    if not qs:
        return qs
    c_type = ContentType.objects.get_for_model(qs.model)
    kwargs = {
        'content_type': c_type,
        'object_id__in': qs.values_list('pk'),
        'level__clearance_level{}'.format(
            '' if exact else '__gt'): clearance_level,
    }
    private_objects = PrivacySetting.objects.filter(**kwargs).values_list(
        'object_id')
    if exact:
        return qs.filter(pk__in=private_objects)
    return qs.exclude(pk__in=private_objects)


class PrivacyLevel(TranslatableModel):
    """
    Privacy level, which defines the assessment factor.

    :name: Translatable name of the privacy level.
    :clearance_level: Privacy level (the higher the more secure).

    """
    translations = TranslatedFields(
        name=models.CharField(max_length=255),
    )

    clearance_level = models.IntegerField(
        unique=True,
        verbose_name=_('Clearance level'),
    )

    class Meta:
        ordering = getattr(
            settings, 'PRIVACY_LEVEL_ORDERING', ['clearance_level'])

    def __unicode__(self):
        return self.safe_translation_getter(
            'name', self.translations.all()[0].name)


class PrivacySetting(models.Model):
    """
    Privacy setting, which defines one custom setting for one object.

    :content_object: Related secured object.
    :level: Connection to the PrivacyLevel model.
    :field_name: Optional field_name to limit the setting to a single field.

    """
    content_type = models.ForeignKey(ContentType)
    object_id = models.PositiveIntegerField()
    content_object = generic.GenericForeignKey('content_type', 'object_id')

    level = models.ForeignKey(
        PrivacyLevel,
        verbose_name=_('Privacy level'),
    )

    field_name = models.CharField(
        max_length=50,
        verbose_name=_('Field name'),
        blank=True,
    )

    def __unicode__(self):
        return '{} - {}'.format(self.content_object, self.level)


privacy/templatetags/privacy_tags.py
"""Template tags of the ``privacy`` app."""
from django.conf import settings
from django.contrib.contenttypes.models import ContentType
from django.core.exceptions import ObjectDoesNotExist
from django.template import Library

from django_libs.loaders import load_member_from_setting

from ..models import PrivacyLevel, PrivacySetting

register = Library()


def get_privacy_dict(obj, field_name):
    """
    Function, which returns a dictionary of relevant data of a privacy level
    (field).

    """
    kwargs = {
        'content_type': ContentType.objects.get_for_model(obj),
        'object_id': obj.pk,
    }
    if field_name:
        kwargs.update({'field_name': field_name})
    else:
        kwargs.update({'field_name': ''})
    privacy_levels = PrivacyLevel.objects.all()
    try:
        selected_setting = PrivacySetting.objects.get(**kwargs)
        selected_level = selected_setting.level.clearance_level
    except PrivacySetting.DoesNotExist:
        selected_setting = None
        selected_level = getattr(
            settings, 'PRIVACY_DEFAULT_CLEARANCE_LEVEL', 1)
    return {
        'field_name': field_name,
        'privacy_levels': privacy_levels,
        'selected_level': selected_level,
        'selected_setting': selected_setting,
    }


@register.assignment_tag
def get_privacy_setting(obj, field_name=None):
    privacy_dict = get_privacy_dict(obj, field_name)
    return privacy_dict['selected_setting']


@register.inclusion_tag('privacy/partials/field.html')
def render_privacy_level_field(obj, field_name=None):
    return get_privacy_dict(obj, field_name)


def is_access_allowed(owner, requester, obj, field_name=None):
    get_clearance_level = load_member_from_setting(
        'PRIVACY_CLEARANCE_LEVEL_FUNCTION')
    clearance_level = get_clearance_level(owner, requester)
    kwargs = {
        'content_type': ContentType.objects.get_for_model(obj),
        'object_id': obj.pk,
    }
    if field_name:
        kwargs.update({'field_name': field_name})
    else:
        kwargs.update({'field_name': ''})
    try:
        privacy_setting = PrivacySetting.objects.get(**kwargs)
    except ObjectDoesNotExist:
        return True  # No settings given? Allow for everyone
    if privacy_setting.level.clearance_level <= clearance_level:
        # Level has to be at least as high as the setting's label of the
        # object/field
        return True
    return False

register.assignment_tag(is_access_allowed)


privacy/migrations/0001_initial.py
# flake8: noqa
# -*- coding: utf-8 -*-
from south.utils import datetime_utils as datetime
from south.db import db
from south.v2 import SchemaMigration
from django.db import models


class Migration(SchemaMigration):

    def forwards(self, orm):
        # Adding model 'PrivacyLevelTranslation'
        db.create_table(u'privacy_privacylevel_translation', (
            (u'id', self.gf('django.db.models.fields.AutoField')(primary_key=True)),
            ('name', self.gf('django.db.models.fields.CharField')(max_length=255)),
            ('language_code', self.gf('django.db.models.fields.CharField')(max_length=15, db_index=True)),
            ('master', self.gf('django.db.models.fields.related.ForeignKey')(related_name='translations', null=True, to=orm['privacy.PrivacyLevel'])),
        ))
        db.send_create_signal(u'privacy', ['PrivacyLevelTranslation'])

        # Adding unique constraint on 'PrivacyLevelTranslation', fields ['language_code', 'master']
        db.create_unique(u'privacy_privacylevel_translation', ['language_code', 'master_id'])

        # Adding model 'PrivacyLevel'
        db.create_table(u'privacy_privacylevel', (
            (u'id', self.gf('django.db.models.fields.AutoField')(primary_key=True)),
            ('clearance_level', self.gf('django.db.models.fields.IntegerField')(unique=True)),
        ))
        db.send_create_signal(u'privacy', ['PrivacyLevel'])

        # Adding model 'PrivacySetting'
        db.create_table(u'privacy_privacysetting', (
            (u'id', self.gf('django.db.models.fields.AutoField')(primary_key=True)),
            ('content_type', self.gf('django.db.models.fields.related.ForeignKey')(to=orm['contenttypes.ContentType'])),
            ('object_id', self.gf('django.db.models.fields.PositiveIntegerField')()),
            ('level', self.gf('django.db.models.fields.related.ForeignKey')(to=orm['privacy.PrivacyLevel'])),
            ('field_name', self.gf('django.db.models.fields.CharField')(max_length=50, blank=True)),
        ))
        db.send_create_signal(u'privacy', ['PrivacySetting'])


    def backwards(self, orm):
        # Removing unique constraint on 'PrivacyLevelTranslation', fields ['language_code', 'master']
        db.delete_unique(u'privacy_privacylevel_translation', ['language_code', 'master_id'])

        # Deleting model 'PrivacyLevelTranslation'
        db.delete_table(u'privacy_privacylevel_translation')

        # Deleting model 'PrivacyLevel'
        db.delete_table(u'privacy_privacylevel')

        # Deleting model 'PrivacySetting'
        db.delete_table(u'privacy_privacysetting')


    models = {
        u'contenttypes.contenttype': {
            'Meta': {'ordering': "('name',)", 'unique_together': "(('app_label', 'model'),)", 'object_name': 'ContentType', 'db_table': "'django_content_type'"},
            'app_label': ('django.db.models.fields.CharField', [], {'max_length': '100'}),
            u'id': ('django.db.models.fields.AutoField', [], {'primary_key': 'True'}),
            'model': ('django.db.models.fields.CharField', [], {'max_length': '100'}),
            'name': ('django.db.models.fields.CharField', [], {'max_length': '100'})
        },
        u'privacy.privacylevel': {
            'Meta': {'ordering': "['clearance_level']", 'object_name': 'PrivacyLevel'},
            'clearance_level': ('django.db.models.fields.IntegerField', [], {'unique': 'True'}),
            u'id': ('django.db.models.fields.AutoField', [], {'primary_key': 'True'})
        },
        u'privacy.privacyleveltranslation': {
            'Meta': {'unique_together': "[('language_code', 'master')]", 'object_name': 'PrivacyLevelTranslation', 'db_table': "u'privacy_privacylevel_translation'"},
            u'id': ('django.db.models.fields.AutoField', [], {'primary_key': 'True'}),
            'language_code': ('django.db.models.fields.CharField', [], {'max_length': '15', 'db_index': 'True'}),
            'master': ('django.db.models.fields.related.ForeignKey', [], {'related_name': "'translations'", 'null': 'True', 'to': u"orm['privacy.PrivacyLevel']"}),
            'name': ('django.db.models.fields.CharField', [], {'max_length': '255'})
        },
        u'privacy.privacysetting': {
            'Meta': {'object_name': 'PrivacySetting'},
            'content_type': ('django.db.models.fields.related.ForeignKey', [], {'to': u"orm['contenttypes.ContentType']"}),
            'field_name': ('django.db.models.fields.CharField', [], {'max_length': '50', 'blank': 'True'}),
            u'id': ('django.db.models.fields.AutoField', [], {'primary_key': 'True'}),
            'level': ('django.db.models.fields.related.ForeignKey', [], {'to': u"orm['privacy.PrivacyLevel']"}),
            'object_id': ('django.db.models.fields.PositiveIntegerField', [], {})
        }
    }

    complete_apps = ['privacy']